modules/core/src/main/java/org/apache/ignite/internal/processors/affinity/GridAffinityProcessor.java

```diff
@@ -212,6 +218,34 @@ public int partition0(String cacheName, Object key, @Nullable AffinityInfo aff)
         return affInfo != null ? F.first(affInfo.assignment().get(partId)) : null;
     }
 
+    /**
+     * @param topVerRmv Collection with affinity topology versions for removing.
+     * @return <tt>true</tt> if this set changed as a result of the call.
+     */
+    public boolean removeCachedAffinity(Collection<AffinityTopologyVersion> topVerRmv) {
+        assert topVerRmv != null;
+
+        Collection<String> caches = ctx.cache().cacheNames();
+        Collection<AffinityAssignmentKey> rmv = new HashSet<>();
+
+        int oldSize = affMap.size();
+
+        for (Map.Entry<AffinityAssignmentKey, IgniteInternalFuture<AffinityInfo>> entry : affMap.entrySet()) {
+            assert entry.getValue() != null;
+
+            if (!entry.getValue().isDone())
+                continue;
+
+            if (topVerRmv.contains(entry.getKey().topVer) || !caches.contains(entry.getKey().cacheName))
+                rmv.add(entry.getKey());
```

Comment:
Why don't you simply use iterator.remove()?

Reply:
fixed